akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventWriter.scala

```diff
@@ -364,29 +365,43 @@ private[akka] object EventWriter {
                 if ((fillSequenceNumberGaps || isSnapshotEvent) && sequenceNumber != 1L) {
                   val reason = if (isSnapshotEvent) AskMaxSeqNrReason.SnapshotEvent else AskMaxSeqNrReason.FillGaps
                   askMaxSeqNr(persistenceId, reason)
-                  StateForPid(waitingForReply = Map.empty, waitingForSeqNrLookup = Vector((repr, replyTo)))
+                  StateForPid(
+                    waitingForReply = Map.empty,
+                    waitingForSeqNrLookup = Vector((repr, replyTo)),
+                    fillSequenceNumberGaps = fillSequenceNumberGaps)
                 } else {
                   sendToJournal(1, Vector(repr))
-                  StateForPid(Map((repr.sequenceNr, (repr, replyTo))), currentTransactionId = 1)
+                  StateForPid(
+                    Map((repr.sequenceNr, (repr, replyTo))),
+                    currentTransactionId = 1,
+                    fillSequenceNumberGaps = fillSequenceNumberGaps)
                 }
 
               case Some(state) =>
+                // For a given persistenceId the fillSequenceNumberGaps is not supposed to change,
+                // but we keep `true` if it has been set to `true`.
+                // isSnapshotEvent and fillSequenceNumberGaps are handled in very similar way, but there is a
+                // difference in the way latest sequence number is cached (see handleUpdatedStateForPid).
+                val newFillSequenceNumberGaps = fillSequenceNumberGaps || state.fillSequenceNumberGaps
                 val expectedSeqNr =
-                  if (fillSequenceNumberGaps || isSnapshotEvent) state.nextExpectedSeqNr
+                  if (newFillSequenceNumberGaps || isSnapshotEvent) state.nextExpectedSeqNr
                   else repr.sequenceNr
 
                 if (state.seqNrlookupInProgress) {
                   context.log.trace2(
                     "Seq nr lookup in progress for persistence id [{}], adding sequence nr [{}] to pending",
                     persistenceId,
                     sequenceNumber)
-                  state.copy(waitingForSeqNrLookup = state.waitingForSeqNrLookup :+ ((repr, replyTo)))
+                  state.copy(
+                    waitingForSeqNrLookup = state.waitingForSeqNrLookup :+ ((repr, replyTo)),
+                    fillSequenceNumberGaps = newFillSequenceNumberGaps || state.fillSequenceNumberGaps)
```

Comment:
Don't need `|| state.fillSequenceNumberGaps` here?

Reply:
right, that was already done in newFillSequenceNumberGaps